pennylane/math/multi_dispatch.py

```diff
@@ -81,6 +83,101 @@ def _multi_dispatch(values):
     return "numpy"
 
 
+def multi_dispatch(argnum=None, tensor_list=None):
+    """Decorater to dispatch arguments handled by the interface.
+
+    This helps simplify definitions of new functions inside PennyLane. We can
+    decorate the function, indicating the arguments that are tensors handled
+    by the interface
+
+
+    >>> @qml.math.multi_dispatch(argnum=[0, 1])
+    ... def some_function(tensor1, tensor2, option, like):
+    ...     # the interface string is stored in `like`.
+    ...     ...
+
+
+    Args:
+        argnum (list[int]): A list of integers indicating indicating the indices
+            to dispatch (i.e., the arguments that are tensors handled by an interface).
+            If ``None``, dispatch over all arguments.
+        tensor_lists (list[int]): a list of integers indicating which indices
+            in ``argnum`` are expected to be lists of tensors.
+            If ``None``, this option is ignored.
+
+    Returns:
+        func: A wrapped version of the function, which will automatically attempt
+        to dispatch to the correct autodifferentiation framework for the requested
+        arguments. Note that the ``like`` argument will be optional, but can be provided
+        if an explicit override is needed.
+
+    .. seealso:: :func:`pennylane.math.multi_dispatch._multi_dispatch`
+
+    .. note::
+        This decorator makes the interface argument "like" optional as it utilizes
+        the utility function `_multi_dispatch` to automatically detect the appropriate
+        interface based on the tensor types.
+
+    **Examples**
+
+    We can redefine external functions to be suitable for PennyLane. Here, we
+    redefine Autoray's ``stack`` function.
+
+    >>> stack = multi_dispatch(argnum=0, tensor_list=0)(autoray.numpy.stack)
+
+    We can also define more elaborate custom function. Here is an example of a ``custom_function`` that
+    computes :math:`c \sum_i (v_i)^T v_i`, where :math:`v_i` are vectors in ``values`` and
+    :math:`c` is a fixed ``coefficient``. Note how ``argnum=0`` only points to the first argument ``values``,
+    how ``tensor_list=0`` indicates that said first argument is a list of vectors, and that ``coefficient`` is not
+    dispatched.
+
+    >>> @math.multi_dispatch(argnum=0,tensor_list=0)
+    >>> def custom_function(values, like, coefficient=10):
+    >>>     # values is a list of vectors
+    >>>     # like can force the interface (optional)
+    >>>     return coefficient * np.sum([math.dot(v,v) for v in values])

```

Comment:
It might be nice to include in the example an example of branching based on `like`, e.g.,

```python
@qml.math.multi_dispatch(argnum=0, tensor_list=0)
def custom_function(values, like, coefficient=10):
    squares = sum([qml.math.dot(v, v) for v in values])

    if like == "tensorflow":
        ...

    return 1j * coefficient * squares
```

just to show what it looks like for developers. Unfortunately, I cannot think of an example that doesn't look very contrived!

Reply:
Thanks for the comments and corrections!

> Unfortunately, I cannot think of an example that doesn't look very contrived!

I also dont know of an example on top of my head. In [math.dot()](https://github.com/PennyLaneAI/pennylane/blob/34d77f88f3e113c9a4824a6708ef1d1a9114c6ff/pennylane/math/multi_dispatch.py#L204) there are exceptions for torch and tensorflow but I think this example may be a bit too convoluted to put there. For now, I just left it with an empty if clause and a comment that this is a possibility in the doc-string.

Reply:
Yes, I also had a look at existing functions to see if there were any that would be nice to use as an example, but all were either too complicated or too specific 😆